include/envoy/network/connection_handler.h

```diff
@@ -68,9 +68,50 @@ class ConnectionHandler {
    * after they have been temporarily disabled.
    */
   virtual void enableListeners() PURE;
+
+  /**
+   * Used by ConnectionHandler to manage listeners.
+   */
+  class ActiveListener {
+  public:
+    virtual ~ActiveListener() = default;
+
+    /**
+     * @return the tag value as configured.
+     */
+    virtual uint64_t listenerTag() PURE;
+    /**
+     * @return the actual Listener object.
+     */
+    virtual ListenerPtr& listener() PURE;
```

Comment:
nit: I would return a raw pointer here vs. a reference to the smart_ptr, or better an optional reference wrapper if it can be null.

Reply:
Actually the caller will call reset() on it...

Reply:
nit: I think it would be better to provide an explicit shutdown(), destroy(), close(), etc. method would be clearer from an interface perspective.

Reply:
I'm not sure I follow. Can't you just have an extra method on this interface called `close()` or `destroy()` or something? Exposing the unique ptr so that the caller can call reset on it seems pretty strange? I think the intent is to just destroy it?

Reply:
I'm not sure I if understand the intention of reset(), but I think we want to destroy the listener but keep the entry in the container? We could expose close() and destroy() and hide listner(), but how to handle reset() call then?

Reply:
I don't follow? Is the only purpose of exposing the listener pointer so that it can be reset/deleted? If so yeah can we just swap it with a close/destroy function?

Reply:
ah, I see what you mean. Yes, `listener().reset()` can be wrapped in a new interface  `destroy()`. However, ConnectionHandlerImpl::findListenerByAddress() expects to return a Listener&, which returns ActiveListener::listener() directly.

Reply:
Right what I'm saying is to have 2 methods, 1 to return the listener which is an optional reference wrapper, and an explicit destroy method. I think that will be much more clear. Thank you!

/wait

Reply:
Got it. I make ActiveListener to inherit from Listener and added destroy() to it. And I also changed listener() to return a pointer to have less change to conn_handler_impl